crates/cext/src/sparse_observable.rs

```diff
@@ -0,0 +1,749 @@
+///
+///   * ``obs`` is a valid, non-null pointer to a ``QkSparseObservable``
+///   * ``cterm`` is a valid, non-null pointer to a ``QkSparseTerm``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_add_term(
+    obs: *mut SparseObservable,
+    cterm: *const CSparseTerm,
+) -> ExitCode {
+    // check pointers are valid and cast them to a reference
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+    let cterm = unsafe { const_ptr_as_ref(cterm) };
+
+    let view = match cterm.try_into() {
+        Ok(view) => view,
+        Err(err) => return ExitCode::from(err),
+    };
+
+    match obs.add_term(view) {
+        Ok(_) => ExitCode::Success,
+        Err(err) => ExitCode::from(err),
+    }
+}
+
+/// @ingroup QkSparseObservable
+/// Get an observable term by reference. This can modify the underlying observable.
+///
+/// @param obs A pointer to the observable.
+/// @param index The index of the term to get.
+/// @param out A pointer to a [CSparseTerm] used to return the observable term.
+///
+/// @return An exit code.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     QkSparseTerm term;
+///     int exit_code = qk_obs_term(obs, 0, &term);
+///     // out-of-bounds indices return an error code
+///     // int error = qk_obs_term(obs, 12, &term);
+///
+/// # Safety
+///
+/// Behavior is undefined if any of the following is violated
+/// * ``obs`` is a valid, non-null pointer to a ``QkSparseObservable``
+/// * ``out`` is a valid, non-null pointer to a ``QkSparseTerm``
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_term(
+    obs: *mut SparseObservable,
+    index: u64,
+    out: *mut CSparseTerm,
+) -> ExitCode {
+    let out = unsafe { mut_ptr_as_ref(out) };
+    let obs = unsafe { mut_ptr_as_ref(obs) };
+
+    let index = index as usize;
+    if index > obs.num_terms() {
+        return ExitCode::IndexError;
+    }
+
+    out.len = obs.boundaries()[index + 1] - obs.boundaries()[index];
+    out.coeff = &mut obs.coeffs_mut()[index];
+    out.num_qubits = obs.num_qubits();
+
+    let start = obs.boundaries()[index];
+    out.bit_terms = &mut obs.bit_terms_mut()[start];
+    out.indices = unsafe { &mut obs.indices_mut()[start] };
+
+    ExitCode::Success
+}
+
+/// @ingroup QkSparseObservable
+/// Get the number of terms in the observable.
+///
+/// @param obs A pointer to the observable.
+///
+/// @return The number of terms in the observable.
+///
+/// # Example
+///
+///     QkSparseObservable *obs = qk_obs_identity(100);
+///     uint64_t num_terms = qk_obs_num_terms(obs);  // num_terms==1
+///
+/// # Safety
+///
+/// Behavior is undefined ``obs`` is not a valid, non-null pointer to a ``QkSparseObservable``.
+#[no_mangle]
+#[cfg(feature = "cbinding")]
+pub unsafe extern "C" fn qk_obs_num_terms(obs: *const SparseObservable) -> u64 {
+    let obs = unsafe { const_ptr_as_ref(obs) };
+    obs.num_terms() as u64
+}
```

Comment:
Why do we cast this one to `u64` when we expose `usize` (as `size_t`) elsewhere?

Reply:
That's legacy code -- initially we tried to avoid `usize` and tried to only use explicitly sized things objects. I'll update